unsafe/string.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package xunsafe
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+// ImmutableBytes represents an immutable byte slice
+type ImmutableBytes []byte
+
+// ToBytes converts a string to a byte slice with zero heap memory allocations.
+// The returned byte slice shares the same memory space as the string passed in.
+// It is the caller's responsibility to make sure the returned byte slice is not
+// modified in any way until end of life.
+func ToBytes(s string) ImmutableBytes {
+	if len(s) == 0 {
+		return nil
+	}
+	sh := (*reflect.StringHeader)(unsafe.Pointer(&s))
+
+	// NB(xichen): it is important that we access s in the same expression
+	// where we assign the data pointer of the string header to the data
+	// pointer of the slice header to make sure the underlying bytes don't
+	// get GC'ed before the assignment happens.
+	return *(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
```

Comment:
> byteHeader.Data is a uintptr so it won't be updated if GC decides to relocate string's memory.

Agreed, but we don't cast the `byteHeader` into a byte slice. We use it to update `b` which is a valid byte slice. Also, this pattern is explicitly marked as a valid pattern in the unsafe package, whereas creating a string or slice through their respective reflect Header structs is explicitly called out as something that one should not do.

> In my example, we also immediately cast SliceHeader to a []byte that uses an unsafe.Pointer instead (https://golang.org/src/runtime/slice.go#L11) which is visible to the GC.

The same is true here since the `unsafe.Pointer` in `b` will be updated by the line `byteHeader.Data = (*reflect.StringHeader)(unsafe.Pointer(&s)).Data`. We are in effect using `byteHeader` to access the private `array` field of a `slice` struct.

>  Memory won't be tracked because type information for the GC is generated by the compiler, and this line of code will generate type information for SliceHeader because this is what we allocate. SliceHeader doesn't have any pointers so the heap area where it ends up will be marked as non-trackable. We cast that memory area to a []byte later, but it doesn't matter.

Agreed, the memory will not be tracked by the `SliceHeader` but it will be tracked by the string `s` which is still live after the call into the `SliceHeader`'s `Data` field.